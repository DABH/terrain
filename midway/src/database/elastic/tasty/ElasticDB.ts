/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2018 Terrain Data, Inc.

import * as Elastic from 'elasticsearch';

import { IsolationLevel, TastyDB, TransactionHandle } from '../../../tasty/TastyDB';
import TastyQuery from '../../../tasty/TastyQuery';
import TastySchema from '../../../tasty/TastySchema';
import TastyTable from '../../../tasty/TastyTable';
import ElasticClient from '../client/ElasticClient';
import ElasticGenerator from './ElasticGenerator';
import ElasticQuery from './ElasticQuery';

export class ElasticDB implements TastyDB
{
  private client: ElasticClient;

  constructor(client: ElasticClient)
  {
    this.client = client;
  }

  public generate(query: TastyQuery)
  {
    return [new ElasticGenerator(query).esQuery];
  }

  public generateString(query: TastyQuery)
  {
    return JSON.stringify(this.generate(query));
  }

  public async schema(skipHidden: boolean = true): Promise<TastySchema>
  {
    const result = await this.client.indices.getMapping({});
    let filteredResponse = result;
    if (skipHidden)
    {
      filteredResponse = this.filterHiddenIndexes(filteredResponse);
    }

    return TastySchema.fromElasticTree(filteredResponse);
  }

  /**
   * Returns the entire ES response object.
   */
  public async query(queries: Elastic.SearchParams[]): Promise<object>
  {
    if (queries.length === 0)
    {
      return [];
    }

    for (let i = 0; ; ++i)
    {
      const result = await this.client.search(queries[i]);

      if (i === queries.length - 1)
      {
        return result;
      }

      await result;
    }
  }

  public async execute(queries: ElasticQuery[])
  {
    let results: object[] = [];
    for (const query of queries)
    {
      let result: any;
      switch (query.op)
      {
        case 'select':
          result = await this.query(query.params as Elastic.SearchParams[]);
          results = results.concat(result.hits.hits);
          break;
        case 'upsert':
          const table: TastyTable = new TastyTable(query.table, query.primaryKeys, query.fields, query.index);
          result = await this.upsert(table, query.params);
          results = results.concat(result);
          break;
        default:
          throw new Error('Unknown query command ' + JSON.stringify(query));
      }
    }

    return results;
  }

  public async destroy()
  {
    // do nothing
  }

  public async storeProcedure(procedure)
  {
    return this.client.putScript(procedure);
  }

  /**
   * Upserts the given objects, based on primary key ('id' in elastic).
   */
  public async upsert(table: TastyTable, elements: object[], handle?: TransactionHandle)
  {
    return this.update_(table, elements, true);
  }

  /**
   * Updates the given objects, based on primary key ('id' in elastic).
   */
  public async update(table: TastyTable, elements: object[], handle?: TransactionHandle)
  {
    return this.update_(table, elements, false);
  }

  public async startTransaction(isolationLevel: IsolationLevel, readOnly: boolean): Promise<TransactionHandle>
  {
    throw new Error('startTransaction() is not supported');
  }

  public async commitTransaction(handle: TransactionHandle): Promise<object[]>
  {
    throw new Error('commitTransaction() is not supported');
  }

  public async rollbackTransaction(handle: TransactionHandle): Promise<object[]>
  {
    throw new Error('rollbackTransaction() is not supported');
  }

  /*
   * Creates the given index
   */
  public async createIndex(index: string)
  {
    return this.client.indices.create({ index });
  }

  /*
   * Refreshes the given index
   */
  public async refreshIndex(index: string | string[])
  {
    return this.client.indices.refresh({ index });
  }

  /*
   * Deletes the given index
   */
  public async deleteIndex(index: string | string[])
  {
    return this.client.indices.delete({ index });
  }

  public async putMapping(table: TastyTable): Promise<object>
  {
    const index = table.getDatabaseName();
    const type = table.getTableName();
    const mapping = table.getMapping();
    return this.putESMapping(index, type, mapping);
  }

  public async putESMapping(index: string, type: string, mapping: object): Promise<object>
  {
    const schema: TastySchema = await this.schema();
    if (schema.databaseNames().indexOf(index) === -1)
    {
      await this.createIndex(index);
    }

    return this.client.indices.putMapping(
      {
        index,
        type,
        body: mapping,
      },
    );
  }

  /*
   * Deletes the given objects based on their primary key
   */
  public async deleteObjects(table: TastyTable, elements: object[])
  {
    const promises: Array<Promise<any>> = [];
    for (const element of elements)
    {
      promises.push(
        this.client.delete(
          {
            id: this.makeID(table, element),
            index: table.getDatabaseName(),
            type: table.getTableName(),
          },
        ),
      );
    }
    return Promise.all(promises);
  }

  /*
   * Refreshes the given index
   */
  public async refresh(index: string)
  {
    return this.client.indices.refresh({ index });
  }

  private async update_(table: TastyTable, elements: object[], upsert: boolean)
  {
    const updated = await this.updateObjects(table, elements, upsert);
    const primaryKeys = table.getPrimaryKeys();
    const results = new Array(updated.length);
    for (let i = 0; i < updated.length; i++)
    {
      results[i] = elements[i];
      if (updated[i]['_id'] !== undefined)
      {
        if ((primaryKeys.length === 1) &&
          (elements[i][primaryKeys[0]] === undefined))
        {
          results[i][primaryKeys[0]] = updated[i]['_id'];
        }
      }
    }
    return results;
  }

  private async updateObjects(table: TastyTable, elements: object[], upsert: boolean)
  {
    if (elements.length > 2)
    {
      await this.bulkUpdate(table, elements, upsert);
      return elements;
    }

    const promises: Array<Promise<any>> = [];
    for (const element of elements)
    {
      let body = {};
      if (upsert)
      {
        body = element;
      }
      else
      {
        body = {
          doc: element,
          doc_as_upsert: true,
        };
      }

      const query = {
        body,
        index: table.getDatabaseName(),
        type: table.getTableName(),
      };

      const compositePrimaryKey = this.makeID(table, element);
      if (compositePrimaryKey !== '')
      {
        query['id'] = compositePrimaryKey;
      }
      promises.push(upsert ? this.client.index(query) : this.client.update(query as any));
    }
    return Promise.all(promises);
  }

  private async bulkUpdate(table: TastyTable, elements: object[], upsert: boolean): Promise<any>
  {
    const body: any[] = [];
    for (const element of elements)
    {
      const cmd = upsert ? 'index' : 'update';
      const command =
        {
          [cmd]: {
            _index: table.getDatabaseName(),
            _type: table.getTableName(),
          },
        };

      const compositePrimaryKey = this.makeID(table, element);
      if (compositePrimaryKey !== '')
      {
        command[cmd]['_id'] = compositePrimaryKey;
      }

      let obj = {};
      if (upsert)
      {
        obj = element;
      }
      else
      {
        obj = {
          doc: element,
          doc_as_upsert: true,
        };
      }

      body.push(command);
      body.push(obj);
    }

    return this.client.bulk({ body });
  }

  private makeID(table: TastyTable, element: object): string
  {
    return table.getPrimaryKeys().map(
      (key: string) =>
      {
        return element[key];
      }).join(table.getPrimaryKeyDelimiter());
  }

  private filterHiddenIndexes(indexes)
  {
    const filteredIndexes = {};

    Object.keys(indexes).forEach((i) =>
    {
      if (!i.startsWith('.'))
      {
        filteredIndexes[i] = indexes[i];
      }
    });

    return filteredIndexes;
  }
}

export default ElasticDB;
