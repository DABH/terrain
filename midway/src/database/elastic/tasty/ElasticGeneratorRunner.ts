/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import * as Elastic from 'elasticsearch';
import TastyNode from '../../../tasty/TastyNode';
import TastyNodeTypes from '../../../tasty/TastyNodeTypes';
import TastyQuery from '../../../tasty/TastyQuery';
import ElasticQuery from './ElasticQuery';
import bodybuilder = require('bodybuilder');

/**
 * Generates elastic queries from TastyQuery objects.
 * Don't use this class directly: use ElasticGenerator instead.
 */
export default class ElasticGeneratorRunner
{
  public esQuery: ElasticQuery;

  constructor(query: TastyQuery)
  {
    this.esQuery = {} as ElasticQuery;
    this.esQuery.index = query.table.getDatabaseName();
    this.esQuery.table = query.table.getTableName();
    this.esQuery.primaryKeys = query.table.getPrimaryKeys();
    this.esQuery.fields = query.table.getColumnNames();
    switch (query.command.tastyType)
    {
      case (TastyNodeTypes.select):
        this.esQuery.op = 'select';
        this.esQuery.params = this.constructSelectQuery(query);
        break;
      case (TastyNodeTypes.upsert):
        this.esQuery.op = 'upsert';
        this.esQuery.params = this.constructUpsertQuery(query);
        break;
      case (TastyNodeTypes.delete):
        this.esQuery.op = 'delete';
        this.esQuery.params = this.constructDeleteQuery(query);
        break;
      default:
        throw new Error('Unknown command in the query:' + query.toString());
    }
  }

  private constructDeleteQuery(query: TastyQuery): object[]
  {
    return this.constructSelectQuery(query);
  }

  private constructUpsertQuery(query: TastyQuery): object[]
  {
    return query.upserts;
  }

  private constructSelectQuery(query: TastyQuery): object[]
  {
    const queryParam: Elastic.SearchParams = {} as Elastic.SearchParams;

    // set table (index) name
    queryParam.index = query.table.getDatabaseName();
    queryParam.type = query.table.getTableName();

    // from clause
    if (query.numSkipped !== 0)
    {
      queryParam.from = query.numSkipped;
    }

    // size clause
    if (query.numTaken !== 0)
    {
      queryParam.size = query.numTaken;
    }

    // start the body
    const body = bodybuilder();

    // stored_fields clause
    if (!query.isSelectingAll())
    {
      const sourceFields: any[] = [];
      for (let i = 0; i < query.selected.length; ++i)
      {
        const column = query.selected[i];
        const columnName = this.getColumnName(column);
        sourceFields.push(columnName);
      }
      body.rawOption('_source', sourceFields);
    }

    if (query.aliases.length !== 0)
    {
      throw new Error('Aliases are not yet supported by ElasticGenerator.');
    }

    if (query.filters.length > 0)
    {
      for (let i = 0; i < query.filters.length; ++i)
      {
        const filter = query.filters[i];
        this.accumulateFilters(body, filter);
      }
    }

    // sort clause
    if (query.sorts.length > 0)
    {
      for (let i = 0; i < query.sorts.length; ++i)
      {
        const sort = query.sorts[i];
        const column = this.getColumnName(sort.node);
        const order = (sort.order === 'asc' ? 'asc' : 'desc');
        body.sort(column, order);
      }
    }
    queryParam['body'] = body.build();
    return [queryParam];
  }

  private accumulateFilters(body: bodybuilder, expression: TastyNode)
  {
    // https://www.elastic.co/guide/en/elasticsearch/guide/current/combining-filters.html#bool-filter
    // currently only supports the basic operators, with the column on the lhs, as well as && and ||

    if (expression.numChildren !== 2)
    {
      throw new Error('Filtering on non-binary expression "' + JSON.stringify(expression) + '".');
    }

    // NB: could be made to accept the column on the rhs too, but currently only supports column on lhs
    const columnName = this.getColumnName(expression.lhs);
    const value = expression.rhs.value; // could be checked for validity

    if (expression.type === '==')
    {
      body.filter('match', columnName, value);
    }
    else if (expression.type === '!=')
    {
      body.notQuery('match', columnName, value);
    }
    else if (expression.type === '<')
    {
      body.filter('range', columnName, { lt: value });
    }
    else if (expression.type === '<=')
    {
      body.filter('range', columnName, { lte: value });
    }
    else if (expression.type === '>')
    {
      body.filter('range', columnName, { gt: value });
    }
    else if (expression.type === '>=')
    {
      body.filter('range', columnName, { gte: value });
    }
    else if (expression.type === '&&')
    {
      const leftQuery = bodybuilder();
      this.accumulateFilters(leftQuery, expression.lhs);
      const rightQuery = bodybuilder();
      this.accumulateFilters(rightQuery, expression.rhs);
      body.andFilter(leftQuery);
      body.andFilter(rightQuery);
    }
    else if (expression.type === '||')
    {
      const leftQuery = bodybuilder();
      this.accumulateFilters(leftQuery, expression.lhs);
      const rightQuery = bodybuilder();
      this.accumulateFilters(rightQuery, expression.rhs);
      body.orFilter(leftQuery);
      body.orFilter(rightQuery);
    }
    else
    {
      throw new Error('Filtering on unsupported expression "' + JSON.stringify(expression) + '".');
    }
  }

  private getColumnName(expression: TastyNode)
  {
    if (expression.type !== '.' || expression.numChildren !== 2)
    {
      throw new Error('Could not find column name in expression "' + JSON.stringify(expression) + '".');
    }

    const table = expression.lhs;
    const column = expression.rhs;

    if (table.type !== 'reference')
    {
      throw new Error('Could not find table name in expression "' + JSON.stringify(expression) + '".');
    }
    if (table.value !== this.esQuery.table)
    {
      throw new Error('Filter expression filters on something other than the queried table "' +
        JSON.stringify(expression) + '".');
    }

    if (column.type !== 'reference')
    {
      throw new Error('Could not find column name in expression "' + JSON.stringify(expression) + '".');
    }

    return column.value;
  }
}
