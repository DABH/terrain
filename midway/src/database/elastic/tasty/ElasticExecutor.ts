/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import * as Elastic from 'elasticsearch';
import TastyExecutor from '../../../tasty/TastyExecutor';
import TastySchema from '../../../tasty/TastySchema';
import TastyTable from '../../../tasty/TastyTable';
import { makePromiseCallback } from '../../../tasty/Utils';
import ElasticClient from '../client/ElasticClient';
import ElasticQuery from './ElasticQuery';

export default class ElasticExecutor implements TastyExecutor
{
  private client: ElasticClient;

  constructor(client: ElasticClient)
  {
    this.client = client;
  }

  /**
   * ES specific extension -- gets the health of the ES cluster
   */
  public async health(): Promise<any>
  {
    return new Promise((resolve, reject) =>
    {
      this.client.cluster.health(
        {},
        makePromiseCallback(resolve, reject));
    });
  }

  public async schema(): Promise<TastySchema>
  {
    const result = await new Promise((resolve, reject) =>
    {
      this.client.indices.getMapping(
        {},
        makePromiseCallback(resolve, reject));
    });

    return TastySchema.fromElasticTree(result);
  }

  /**
   * Returns the entire ES response object.
   */
  public async fullQuery(queryObject: Elastic.SearchParams): Promise<any>
  {
    return new Promise((resolve, reject) =>
    {
      this.client.search(
        queryObject,
        makePromiseCallback(resolve, reject));
    });
  }

  /**
   * returns only the query hits
   */
  public async query(queryObject: Elastic.SearchParams)
  {
    const result = await this.fullQuery(queryObject);
    return result.hits.hits;
  }

  public async destroy()
  {
    // do nothing
  }

  public storeProcedure(procedure)
  {
    return new Promise(
      (resolve, reject) =>
      {
        this.client.putScript(
          procedure,
          makePromiseCallback(resolve, reject));
      });
  }

  /**
   * Upserts the given objects, based on primary key ('id' in elastic).
   */
  public async upsertObjects(table: TastyTable, elements: object[])
  {
    if (elements.length > 2)
    {
      return this.bulkUpsert(table, elements);
    }

    const promises = [];

    for (const element of elements)
    {
      promises.push(
        new Promise((resolve, reject) =>
        {
          const query = {
            body: element,
            id: this.makeID(table, element),
            index: table.getDatabaseName(),
            type: table.getTableName(),
          };

          this.client.index(
            query,
            makePromiseCallback(resolve, reject));
        }),
      );
    }
    await Promise.all(promises);
  }

  /*
   * Deletes the given objects based on their primary key
   */
  public async deleteIndex(indexName)
  {
    return new Promise((resolve, reject) =>
    {
      const params = {
        index: indexName,
      };

      this.client.indices.delete(
        params,
        makePromiseCallback(resolve, reject));
    });
  }

  /*
   * Deletes the given objects based on their primary key
   */
  public async deleteDocumentsByID(table: TastyTable, elements: object[])
  {
    const promises = [];
    for (const element of elements)
    {
      promises.push(
        new Promise((resolve, reject) =>
        {
          const params =
            {
              id: this.makeID(table, element),
              index: table.getDatabaseName(),
              type: table.getTableName(),
            };

          this.client.delete(
            params,
            makePromiseCallback(resolve, reject));
        }));
    }
    await Promise.all(promises);
  }

  public async executeElasticTastyQuery(query: ElasticQuery)
  {
    switch (query.op)
    {
      case 'select':
        return await this.executeElasticTastySelectQuery(query);
      default:
        throw new Error('Unknown query command ' + query);
    }
  }

  private async executeElasticTastySelectQuery(query: ElasticQuery)
  {
    return await this.query(query.param as Elastic.SearchParams);
  }

  private async bulkUpsert(table: TastyTable, elements: object[]): Promise<any>
  {
    const body = [];
    for (let i = 0; i < elements.length; ++i)
    {
      const element = elements[i];
      const command =
        {
          index: {
            _id: this.makeID(table, element),
            _index: table.getDatabaseName(),
            _type: table.getTableName(),
          },
        };

      body.push(command);
      body.push(element);
    }

    return new Promise((resolve, reject) =>
    {
      this.client.bulk(
        {
          body,
        },
        makePromiseCallback(resolve, reject));
    });
  }

  private makeID(table: TastyTable, element: object): string
  {
    return table.getPrimaryKeys(element).map(
      (key: string) =>
      {
        return element[key];
      }).join('-');
  }
}
