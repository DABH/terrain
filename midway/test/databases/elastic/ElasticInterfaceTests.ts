/*
University of Illinois/NCSA Open Source License 

Copyright (c) 2018 Terrain Data, Inc. and the authors. All rights reserved.

Developed by: Terrain Data, Inc. and
              the individuals who committed the code in this file.
              https://github.com/terraindata/terrain
                  
Permission is hereby granted, free of charge, to any person 
obtaining a copy of this software and associated documentation files 
(the "Software"), to deal with the Software without restriction, 
including without limitation the rights to use, copy, modify, merge,
publish, distribute, sublicense, and/or sell copies of the Software, 
and to permit persons to whom the Software is furnished to do so, 
subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, 
  this list of conditions and the following disclaimers.

* Redistributions in binary form must reproduce the above copyright 
  notice, this list of conditions and the following disclaimers in the 
  documentation and/or other materials provided with the distribution.

* Neither the names of Terrain Data, Inc., Terrain, nor the names of its 
  contributors may be used to endorse or promote products derived from
  this Software without specific prior written permission.

This license supersedes any copyright notice, license, or related statement
following this comment block.  All files in this repository are provided
under the same license, regardless of whether a corresponding comment block
appears in them.  This license also applies retroactively to any previous
state of the repository, including different branches and commits, which
were made public on or after December 8th, 2018.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS WITH
THE SOFTWARE.
*/

// Copyright 2017 Terrain Data, Inc.

import * as fs from 'fs';
import * as hash from 'object-hash';
import * as winston from 'winston';
import { makePromiseCallback } from '../../../src/tasty/Utils';

import ElasticInterface from '../../../src/databases/elastic/ElasticInterface';

let elasticInterface;

beforeAll(() =>
{
  elasticInterface = new ElasticInterface();
});

test('elastic health', async (done) =>
{
  const result = await new Promise((resolve, reject) =>
  {
    elasticInterface.cluster.health(
      {},
      makePromiseCallback(resolve, reject));
  });
  winston.info(result);
  done();
});

test('basic query', async (done) =>
{
  try
  {
    const result = await new Promise((resolve, reject) =>
    {
      elasticInterface.search(
        {
          index: 'movies',
          type: 'data',
          body: {
            query: {},
            sort: [{ revenue: 'desc' }, { movieid: 'asc' }],
          },
          size: 1,
        },
        makePromiseCallback(resolve, reject));
    });
    winston.info(JSON.stringify(result, null, 2));
    expect(result.hits).toEqual({
      total: 27278,
      max_score: null,
      hits: [
        {
          _index: 'movies',
          _type: 'data',
          _id: '72998',
          _score: null,
          _source: {
            'overview': 'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.',
            'votecount': 8218,
            'posterpath': '/tcqb9NHdw9SWs2a88KCDD4V8sVR.jpg',
            'runtime': 162,
            'movieid': 72998,
            'language': 'en',
            'releasedate': '2009-12-10T00:00:00.000Z',
            'voteaverage': 7.099999904632568,
            'title': 'Avatar (2009)',
            'revenue': 2100000000,
            'backdroppath': '/5XPPB44RQGfkBrbJxmtdndKz05n.jpg',
            'genres': 'Action|Adventure|Sci-Fi|IMAX',
            'popularity': 8.273819923400879,
            '@version': '1',
            'tagline': 'Enter the World of Pandora.',
            'status': 'Released',
            'budget': 237000000,
            'homepage': 'http://www.avatarmovie.com/',
          },
          sort: [
            2100000000,
            72998,
          ],
        },
      ],
    });
  }
  catch (e)
  {
    fail(e);
  }

  done();
});

test('put script', async (done) =>
{
  try
  {
    await new Promise((resolve, reject) =>
    {
      elasticInterface.putScript(
        {
          id: 'terrain_test_movie_profit',
          lang: 'painless',
          body: {
            script: `return doc['revenue'].value - doc['budget'].value;`,
          },
        },
        makePromiseCallback(resolve, reject));
    });

    const result = await new Promise((resolve, reject) =>
    {
      elasticInterface.search(
        {
          index: 'movies',
          type: 'data',
          body: {
            query: {},
            sort: {
              _script: {
                type: 'number',
                order: 'desc',
                script: {
                  stored: 'terrain_test_movie_profit',
                  params: {},
                },
              },
            },
          },
          size: 1,
        },
        makePromiseCallback(resolve, reject));
    });
    winston.info(JSON.stringify(result, null, 2));
    expect(result.hits).toEqual({
      total: 27278,
      max_score: null,
      hits: [
        {
          _index: 'movies',
          _type: 'data',
          _id: '72998',
          _score: null,
          _source: {
            'overview': 'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.',
            'votecount': 8218,
            'posterpath': '/tcqb9NHdw9SWs2a88KCDD4V8sVR.jpg',
            'runtime': 162,
            'movieid': 72998,
            'language': 'en',
            'releasedate': '2009-12-10T00:00:00.000Z',
            'voteaverage': 7.099999904632568,
            'title': 'Avatar (2009)',
            'revenue': 2100000000,
            'backdroppath': '/5XPPB44RQGfkBrbJxmtdndKz05n.jpg',
            'genres': 'Action|Adventure|Sci-Fi|IMAX',
            'popularity': 8.273819923400879,
            '@version': '1',
            'tagline': 'Enter the World of Pandora.',
            'status': 'Released',
            'budget': 237000000,
            'homepage': 'http://www.avatarmovie.com/',
          },
          sort: [
            1863000000,
          ],
        },
      ],
    });
  }
  catch (e)
  {
    fail(e);
  }

  done();
});
