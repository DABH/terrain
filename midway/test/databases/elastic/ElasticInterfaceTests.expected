{
  "basic query" : {
    "total" : 27278,
    "max_score" : null,
    "hits" : [
      {
        "_index" : "movies",
        "_type" : "data",
        "_id" : "72998",
        "_score" : null,
        "_source" : {
          "overview" : "In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.",
          "votecount" : 8218,
          "posterpath" : "/tcqb9NHdw9SWs2a88KCDD4V8sVR.jpg",
          "runtime" : 162,
          "movieid" : 72998,
          "language" : "en",
          "releasedate" : "2009-12-10T00:00:00.000Z",
          "voteaverage" : 7.099999904632568,
          "title" : "Avatar (2009)",
          "revenue" : 2100000000,
          "backdroppath" : "/5XPPB44RQGfkBrbJxmtdndKz05n.jpg",
          "genres" : "Action|Adventure|Sci-Fi|IMAX",
          "popularity" : 8.273819923400879,
          "@version" : "1",
          "tagline" : "Enter the World of Pandora.",
          "status" : "Released",
          "budget" : 237000000,
          "homepage" : "http://www.avatarmovie.com/"
        },
        "sort" : [
          2100000000,
          72998
        ]
      }
    ]
  },
  "put script" : {
    "total" : 27278,
    "max_score" : null,
    "hits" : [
      {
        "_index" : "movies",
        "_type" : "data",
        "_id" : "72998",
        "_score" : null,
        "_source" : {
          "overview" : "In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.",
          "votecount" : 8218,
          "posterpath" : "/tcqb9NHdw9SWs2a88KCDD4V8sVR.jpg",
          "runtime" : 162,
          "movieid" : 72998,
          "language" : "en",
          "releasedate" : "2009-12-10T00:00:00.000Z",
          "voteaverage" : 7.099999904632568,
          "title" : "Avatar (2009)",
          "revenue" : 2100000000,
          "backdroppath" : "/5XPPB44RQGfkBrbJxmtdndKz05n.jpg",
          "genres" : "Action|Adventure|Sci-Fi|IMAX",
          "popularity" : 8.273819923400879,
          "@version" : "1",
          "tagline" : "Enter the World of Pandora.",
          "status" : "Released",
          "budget" : 237000000,
          "homepage" : "http://www.avatarmovie.com/"
        },
        "sort" : [
          1863000000
        ]
      }
    ]
  }
}
